File: backend/src/MatchLogic.Domain/DataProfiling/ColumnProfile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MatchLogic.Domain.DataProfiling
{
    public class ColumnProfile
    {
        public string FieldName { get; set; }
        public string Type { get; set; }
        public int Length { get; set; }
        public string Pattern { get; set; }

        // Count statistics
        public long Total { get; set; }
        public long Valid { get; set; }
        public long Invalid { get; set; }
        public long Filled { get; set; }
        public long Null { get; set; }
        public long Distinct { get; set; }

        // Character counts
        public long Numbers { get; set; }
        public long NumbersOnly { get; set; }
        public long Letters { get; set; }
        public long LettersOnly { get; set; }
        public long NumbersAndLetters { get; set; }
        public long Punctuation { get; set; }
        public long LeadingSpaces { get; set; }
        public long NonPrintableCharacters { get; set; }

        // Min/Max/Mean/Median/Mode
        public string Min { get; set; }
        public string Max { get; set; }
        public string Mean { get; set; }
        public string Median { get; set; }
        public string Mode { get; set; }
        public string Extreme { get; set; }

        // Sample values for UI display
        public List<string> SampleValues { get; set; } = new();

        // Value distribution for categorical fields
        public Dictionary<string, long> ValueDistribution { get; set; }

        public List<PatternInfo> Patterns { get; set; }

        // Store document IDs for each type of row reference
        public Dictionary<ProfileCharacteristic, Guid> CharacteristicRowDocumentIds { get; set; } = new();
        public Dictionary<string, Guid> PatternMatchRowDocumentIds { get; set; } = new();
        public Dictionary<string, Guid> ValueRowDocumentIds { get; set; } = new();
    }

}

File: frontend/src/modules/DataProfiling/models/column-profile.ts
export interface ColumnProfile {
  [key: string]: DataProfile;
}

export interface DataProfile {
    inferredDataType: string;
    typeDetectionConfidence: number;
    typeDetectionResults: TypeDetectionResult[];
    histogram: any;
    outliers: any[];
    clusters: Cluster[];
    skewness: number;
    kurtosis: number;
    interquartileRange: any;
    qualityScore: QualityScore;
    discoveredPatterns: any[];
    appliedRules: Rule[];
    violations: Violation[];
    detectedFormat: DetectedFormat;
    possibleSemanticTypes: SemanticType[];
    warnings: any[];
    fieldName: string;
    type: string;
    length: number;
    pattern: string;
    total: number;
    valid: number;
    invalid: number;
    filled: number;
    null: number;
    distinct: number;
    numbers: number;
    numbersOnly: number;
    letters: number;
    lettersOnly: number;
    numbersAndLetters: number;
    punctuation: number;
    leadingSpaces: number;
    nonPrintableCharacters: number;
    min: string;
    max: string;
    mean: any;
    median: any;
    mode: any;
    extreme: any;
    sampleValues: any[];
    valueDistribution: any;
    patterns: Pattern[];
    characteristicRowDocumentIds: Record<string, any>;
    patternMatchRowDocumentIds: Record<string, any>;
    valueRowDocumentIds: Record<string, any>;
}

interface TypeDetectionResult {
  dataType: string;
  confidence: number;
}

interface Cluster {
  clusterId: number;
  count: number;
  centroid: number;
  representative: string | null;
  sampleValues: string[];
  sampleRows: SampleRow[];
}

interface SampleRow {
  rowData: RowData | null;
  value: string;
  rowNumber: number;
}

export interface RowData {
  _id: {
    timestamp: number;
    machine: number;
    pid: number;
    increment: number;
    creationTime: string;
  };
  _metadata: {
    RowNumber: number;
    Hash: string;
    SourceFile: string;
    BlockingKey: string | null;
  };
  [key: string]: string | number | boolean | object | null | undefined;
}

interface QualityScore {
  overallScore: number;
  completeness: number;
  accuracy: number;
  consistency: number;
  uniqueness: number;
  validity: number;
}

interface Rule {
  ruleName: string;
  description: string;
  passCount: number;
  failCount: number;
  parameters: Record<string, string>;
}

interface Violation {
  ruleName: string;
  value: string | null;
  message: string;
  examples: SampleRow[];
}

interface DetectedFormat {
  format: string;
  confidence: number;
  examples: any[];
}

export interface SemanticType {
  type: string;
  confidence: number;
}

interface Pattern {
  pattern: string;
  count: number;
  matchPercentage: number;
}
